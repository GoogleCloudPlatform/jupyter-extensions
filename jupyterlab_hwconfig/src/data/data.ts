/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MachineType, MachineTypeConfiguration } from './machine_types';
import {
  Accelerator,
  nvidiaNameToEnum,
  NO_ACCELERATOR_COUNT,
} from './accelerator_types';

export interface Option {
  text: string;
  value: string | number;
  disabled?: boolean;
}

interface Instance {
  attributes: {
    framework: string;
    title: string;
    version: string;
  };
  cpuPlatform: string;
  id: number;
  image: string;
  machineType: MachineType;
  name: string;
  zone: string;
}

interface Project {
  numericProjectId: number;
  projectId: string;
}

export interface Utilization {
  cpu: number;
  memory: number;
}

interface Gpu {
  name: string;
  driver_version: string;
  cuda_version: string;
  count: string;
  gpu: number;
  memory: number;
  temperature: number;
}

export interface Details {
  instance: Instance;
  project: Project;
  utilization: Utilization;
  gpu: Gpu;
  acceleratorTypes: Accelerator[];
  machineTypes: MachineTypeConfiguration[];
}

export interface HardwareConfiguration {
  machineType: MachineType;
  attachGpu: boolean;
  gpuType: string; // as Notebooks API AcceleratorType enum values
  gpuCount: string;
  diskType?: string;
  diskSizeGb?: string;
}

export function isEqualHardwareConfiguration(
  a: HardwareConfiguration,
  b: HardwareConfiguration
) {
  return (
    a.machineType.name === b.machineType.name &&
    a.machineType.description === b.machineType.description &&
    a.attachGpu === b.attachGpu &&
    a.gpuType === b.gpuType &&
    a.gpuCount === b.gpuCount
  );
}

export function detailsToHardwareConfiguration(
  details: Details
): HardwareConfiguration {
  const { instance, gpu } = details;

  return {
    machineType: instance.machineType,
    attachGpu: Boolean(gpu.name),
    gpuType: nvidiaNameToEnum(gpu.name),
    gpuCount: gpu.name ? gpu.count : NO_ACCELERATOR_COUNT,
  };
}

/** The average number of hours in a month, i.e., 365 / 12 * 24 */
export const HOURS_PER_MONTH = 730;
export const NO_SUSTAINED_DISCOUNT_PREFIXES = ['e2-'];

export const REGIONS: Option[] = [
  {
    value: 'us-central1',
    text: 'us-central1 (Iowa)',
  },
  {
    value: 'us-east1',
    text: 'us-east1 (South Carolina)',
  },
  {
    value: 'us-east4',
    text: 'us-east4 (Northern Virginia)',
  },
  {
    value: 'us-west1',
    text: 'us-west1 (Oregon)',
  },
  {
    value: 'us-west2',
    text: 'us-west2 (Los Angeles)',
  },
  {
    value: 'us-west3',
    text: 'us-west3 (Salt Lake City)',
  },
  {
    value: 'asia-east1',
    text: 'asia-east1 (Taiwan)',
  },
  {
    value: 'europe-north1',
    text: 'europe-north1 (Finland)',
  },
  {
    value: 'europe-west1',
    text: 'europe-west1 (Belgium)',
  },
  {
    value: 'europe-west2',
    text: 'europe-west2 (London)',
  },
  {
    value: 'europe-west3',
    text: 'europe-west3 (Frankfurt)',
  },
  {
    value: 'europe-west4',
    text: 'europe-west4 (Netherlands)',
  },
  {
    value: 'europe-west6',
    text: 'europe-west6 (Zurich)',
  },
  {
    value: 'asia-east1',
    text: 'asia-east1 (Taiwan)',
  },
  {
    value: 'asia-east2',
    text: 'asia-east2 (Hong Kong)',
  },
  {
    value: 'asia-south1',
    text: 'asia-south1 (Mumbai)',
  },
  {
    value: 'asia-northeast1',
    text: 'asia-northeast1 (Tokyo)',
  },
  {
    value: 'asia-northeast2',
    text: 'asia-northeast2 (Osaka)',
  },
  {
    value: 'asia-northeast3',
    text: 'asia-northeast3 (Seoul)',
  },
  {
    value: 'asia-southeast1',
    text: 'asia-southeast1 (Singapore)',
  },
];

export function extractLast(str: string) {
  const last = str.split('/').pop();
  return last;
}

interface AttributeMapper {
  label: string;
  mapper: (details: Details) => string;
}

// Displayable virtual machine detail attributes
export const MAPPED_ATTRIBUTES: AttributeMapper[] = [
  { label: 'VM Name', mapper: (details: Details) => details.instance.name },
  { label: 'Project', mapper: (details: Details) => details.project.projectId },
  {
    label: 'Framework',
    mapper: (details: Details) => details.instance.attributes.framework,
  },
  {
    label: 'Machine Type',
    mapper: (details: Details) =>
      `${details.instance.machineType.description} (${details.instance.machineType.name})`,
  },
];

export const REFRESHABLE_MAPPED_ATTRIBUTES = [
  {
    label: 'CPU Utilization',
    mapper: (details: Details) => `CPU: ${details.utilization.cpu.toFixed(1)}%`,
  },
  {
    label: 'Memory Utilization',
    mapper: (details: Details) =>
      `Memory: ${details.utilization.memory.toFixed(1)}%`,
  },
  {
    label: 'GPU Utilization',
    mapper: (details: Details) => {
      if (!details.gpu.name) {
        return 'No GPUs';
      }
      return `GPU: ${details.gpu.name} - ${details.gpu.gpu.toFixed(1)}%`;
    },
  },
];

MAPPED_ATTRIBUTES.push(...REFRESHABLE_MAPPED_ATTRIBUTES);
