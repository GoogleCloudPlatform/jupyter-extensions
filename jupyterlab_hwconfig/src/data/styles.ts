/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import { stylesheet } from 'typestyle';
import { BASE_FONT } from 'gcp_jupyterlab_shared';

/* Class names applied to the component. */
export const STYLES = stylesheet({
  container: {
    color: 'var(--jp-ui-font-color1)',
    fontFamily: 'var(--jp-ui-font-family)',
    fontSize: 'var(--jp-ui-font-size1, 13px)',
    lineHeight: '24px',
    alignItems: 'center',
    ...csstips.horizontal,
  },
  containerPadding: {
    margin: '0px',
    padding: '35px 30px',
    backgroundColor: 'var(--jp-layout-color1)',
  },
  containerSize: {
    width: 468,
  },
  heading: {
    fontSize: '22px',
    paddingBottom: '10px',
    fontWeight: 400,
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
    display: 'block',
  },
  subheading: {
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
    fontWeight: 700,
    fontSize: '15px',
    paddingTop: '20px',
    paddingBottom: '5px',
    display: 'block',
  },
  paragraph: {
    fontSize: '14px',
    paddingTop: '2px',
    display: 'block',
    fontWeight: 400,
    fontFamily: 'var(--jp-ui-font-family)',
    color: 'var(--jp-ui-font-color1)',
  },
  attribute: {
    marginRight: '4px',
  },
  interactiveHover: {
    $nest: {
      '&:hover': {
        backgroundColor: '#8a8a8a',
      },
    },
  },
  icon: {
    display: 'inline-block',
    height: '18px',
    marginRight: '4px',
    width: '18px',
  },
  listRow: {
    display: 'table-row',
    boxShadow: 'inset 0 -1px 0 0 var(--jp-border-color1)',
  },
  infoMessage: {
    marginTop: '20px',
  },
});

export const TEXT_STYLE = {
  fontFamily: BASE_FONT.fontFamily as string,
  fontSize: BASE_FONT.fontSize as number,
  color: 'var(--jp-ui-font-color1)',
};
