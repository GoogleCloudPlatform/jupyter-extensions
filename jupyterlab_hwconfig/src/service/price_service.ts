/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ServerConnection } from '@jupyterlab/services';
import {
  HardwareConfiguration,
  REGIONS,
  HOURS_PER_MONTH,
  NO_SUSTAINED_DISCOUNT_PREFIXES,
} from '../data/data';

/** The monthly price catalog for region and sku */
type Catalog = Map<string, Map<string, number>>;

export class PriceService {
  private catalog: Catalog;

  // Exposed for testing purposes only
  set _catalog(catalog: Catalog) {
    this.catalog = catalog;
  }

  private async buildCatalog(): Promise<void> {
    // Only attempt to build the catalog once
    if (this.catalog) {
      return;
    }
    try {
      this.catalog = new Map();
      const response = await ServerConnection.makeRequest(
        `${ServerConnection.defaultSettings.baseUrl}gcp/v1/pricelist`,
        {},
        ServerConnection.defaultSettings
      );
      if (!response.ok) {
        throw { result: await response.text() };
      }
      const fullCatalog = await response.json();
      const pricelist = fullCatalog.gcp_price_list;
      Object.keys(pricelist).forEach(sku => {
        if (sku.endsWith('PREEMPTIBLE')) return;
        if (
          sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-N1') ||
          sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-E2')
        ) {
          this.populateCatalog(
            sku.substring(25).toLowerCase(),
            pricelist[sku],
            true
          );
        } else if (sku.startsWith('GPU_')) {
          this.populateCatalog(sku.substring(4), pricelist[sku], true);
        } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-CAPACITY') {
          this.populateCatalog('pd-standard', pricelist[sku], false);
        } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-SSD') {
          this.populateCatalog('pd-ssd', pricelist[sku], false);
        }
      });
    } catch (err) {
      console.error(`Unable to retrieve pricelist.`);
    }
  }

  // tslint:disable-next-line:no-any The json object returned is not typed.
  private populateCatalog(sku: string, prices: any, timeBased: boolean) {
    Object.keys(prices).forEach(region => {
      const regionCatalog = this.getRegionCatalog(region);
      if (regionCatalog) {
        regionCatalog.set(
          sku,
          prices[region] * (timeBased ? HOURS_PER_MONTH : 1.0)
        );
      }
    });
  }

  private getRegionCatalog(region: string): Map<string, number> | undefined {
    if (REGIONS.find(r => r.value === region)) {
      if (!this.catalog.has(region)) {
        this.catalog.set(region, new Map<string, number>());
      }
      return this.catalog.get(region);
    }
    return undefined;
  }

  private getRegionFromZone(zone: string): string {
    return zone.split('-', 2).join('-');
  }

  /**
   * Returns estimated total monthly price for a given instance to be created in a
   * certain region. If catalog doesn't contain relevant resource information
   * returns undefined.
   */
  async getPrice(
    zone: string,
    config: HardwareConfiguration
  ): Promise<number | undefined> {
    await this.buildCatalog();

    const regionCatalog = this.getRegionCatalog(this.getRegionFromZone(zone!));
    if (!regionCatalog) return undefined;

    let total = 0;
    let hasSustainedUseDiscount = true;

    if (config.machineType) {
      const machineTypeValue = config.machineType.name;
      total += regionCatalog.get(machineTypeValue);

      hasSustainedUseDiscount = !NO_SUSTAINED_DISCOUNT_PREFIXES.some(prefix =>
        machineTypeValue.startsWith(prefix)
      );
    }

    if (config.attachGpu && config.gpuType && config.gpuCount) {
      total += regionCatalog.get(config.gpuType) * parseInt(config.gpuCount);
    }

    if (config.diskType && config.diskSizeGb) {
      total += regionCatalog.get(config.diskType) * parseInt(config.diskSizeGb);
    }

    if (hasSustainedUseDiscount) {
      total *= 0.7;
    }

    return isNaN(total) ? undefined : total;
  }
}
