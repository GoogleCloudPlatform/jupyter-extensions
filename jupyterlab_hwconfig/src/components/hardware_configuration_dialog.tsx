/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { Dialog } from '@material-ui/core';
import { HardwareScalingForm } from './hardware_scaling_form';
import { HardwareScalingStatus } from './hardware_scaling_status';
import { ConfirmationPage } from './confirmation_page';
import { DetailsDialogBody } from './details_dialog_body';
import {
  HardwareConfiguration,
  DetailsResponse,
  detailsToHardwareConfiguration,
} from '../data/data';
import { authTokenRetrieval } from '../service/auth_token_retrieval';
import { HardwareService } from '../service/hardware_service';
import { NotebooksService } from '../service/notebooks_service';
import { PriceService } from '../service/price_service';
import { Accelerator } from '../data/accelerator_types';
import {
  MachineTypeConfiguration,
  getMachineTypeConfigurations,
} from '../data/machine_types';

export enum View {
  DETAILS,
  FORM,
  CONFIRMATION,
  STATUS,
}

interface Props {
  open: boolean;
  receivedError: boolean;
  hardwareService: HardwareService;
  notebookService: NotebooksService;
  priceService: PriceService;
  onClose: () => void;
  onCompletion: () => void;
  details?: DetailsResponse;
}

interface State {
  view: View;
  hardwareConfiguration: HardwareConfiguration;
  acceleratorTypes?: Accelerator[];
  machineTypes?: MachineTypeConfiguration[];
}

export class HardwareConfigurationDialog extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      view: View.DETAILS,
      hardwareConfiguration: null,
    };
  }

  render() {
    return <Dialog open={this.props.open}>{this.getDisplay()}</Dialog>;
  }

  private getDisplay() {
    const {
      details,
      receivedError,
      hardwareService,
      notebookService,
      priceService,
      onClose,
      onCompletion,
    } = this.props;
    const {
      acceleratorTypes,
      hardwareConfiguration,
      machineTypes,
      view,
    } = this.state;

    switch (view) {
      case View.DETAILS:
        return (
          <DetailsDialogBody
            details={details}
            receivedError={receivedError}
            onDialogClose={onClose}
            onUpdate={() => this.onDisplayForm()}
          />
        );

      case View.FORM:
        return (
          <HardwareScalingForm
            acceleratorTypes={acceleratorTypes}
            details={details}
            machineTypes={machineTypes}
            priceService={priceService}
            onDialogClose={onClose}
            onSubmit={(config: HardwareConfiguration) => {
              this.setState({
                view: View.CONFIRMATION,
                hardwareConfiguration: config,
              });
            }}
          />
        );

      case View.CONFIRMATION:
        return (
          <ConfirmationPage
            formData={hardwareConfiguration}
            currentConfiguration={detailsToHardwareConfiguration(details)}
            onDialogClose={onClose}
            onSubmit={() => {
              this.setState({
                view: View.STATUS,
              });
            }}
          />
        );

      case View.STATUS:
        return (
          <HardwareScalingStatus
            authTokenRetrieval={authTokenRetrieval}
            machineTypes={machineTypes}
            hardwareConfiguration={hardwareConfiguration}
            hardwareService={hardwareService}
            notebookService={notebookService}
            onDialogClose={onClose}
            onCompletion={onCompletion}
          />
        );
    }
  }

  private async onDisplayForm() {
    const { hardwareService } = this.props;
    let { acceleratorTypes, machineTypes } = this.state;
    if (!(acceleratorTypes?.length && machineTypes?.length)) {
      const [
        machineTypesResponse,
        acceleratorTypesResponse,
      ] = await Promise.all([
        hardwareService.getMachineTypes(),
        hardwareService.getAcceleratorTypes(),
      ]);

      machineTypes = getMachineTypeConfigurations(machineTypesResponse);
      acceleratorTypes = acceleratorTypesResponse;
    }

    this.setState({
      machineTypes,
      acceleratorTypes,
      view: View.FORM,
    });
  }
}
