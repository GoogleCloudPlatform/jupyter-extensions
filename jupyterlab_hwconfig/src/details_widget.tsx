/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { ReactWidget } from '@jupyterlab/apputils';
import { classes } from 'typestyle';
import { STYLES } from './data/styles';
import {
  Details,
  MAPPED_ATTRIBUTES,
  REFRESHABLE_MAPPED_ATTRIBUTES,
} from './data/data';
import { getMachineTypeConfigurations } from './data/machine_types';
import { HardwareConfigurationDialog } from './components/hardware_configuration_dialog';
import { ResourceUtilizationCharts } from './components/resource_utilization_charts';
import { ServerWrapper } from './components/server_wrapper';
import { WidgetPopup } from './components/widget_popup';
import { NotebooksService } from './service/notebooks_service';
import { DetailsService } from './service/details_service';
import { PriceService } from './service/price_service';

interface Props {
  detailsServer: ServerWrapper;
  detailsService: DetailsService;
  notebookService: NotebooksService;
  priceService: PriceService;
  initializationError?: boolean;
}

interface State {
  receivedError: boolean;
  shouldRefresh: boolean;
  dialogDisplayed: boolean;
  displayedAttributes: [number, number];
  details?: Details;
}

const ICON_CLASS = 'jp-VmStatusIcon';

/** Instance details display widget */
export class VmDetails extends React.Component<Props, State> {
  private readonly refreshInterval: number;

  constructor(props: Props) {
    super(props);
    this.state = {
      displayedAttributes: [0, 1],
      receivedError: props.initializationError,
      shouldRefresh: false,
      dialogDisplayed: false,
    };
    this.refreshInterval = window.setInterval(() => {
      if (this.state.shouldRefresh) {
        this.updateUtilizationData();
      }
    }, 1500);
  }

  componentDidMount() {
    const { initializationError } = this.props;
    !initializationError && this.initializeDetailsFromServer();
  }

  componentWillUnmount() {
    window.clearInterval(this.refreshInterval);
  }

  render() {
    const { details, dialogDisplayed, receivedError } = this.state;
    const { detailsServer, notebookService, priceService } = this.props;
    const noDetailsMessage = receivedError
      ? 'Error retrieving VM Details'
      : 'Retrieving VM Details...';

    return (
      <span className={STYLES.container}>
        <span
          className={classes(STYLES.icon, ICON_CLASS, STYLES.interactiveHover)}
          title="Show all details"
          onClick={() => this.setState({ dialogDisplayed: true })}
        ></span>
        <WidgetPopup>
          <ResourceUtilizationCharts detailsServer={detailsServer} />
        </WidgetPopup>
        <span className={STYLES.interactiveHover}>
          {details ? this.getDisplayedDetails(details) : noDetailsMessage}
        </span>
        {dialogDisplayed && (
          <HardwareConfigurationDialog
            open={dialogDisplayed}
            details={details}
            detailsServer={detailsServer}
            receivedError={receivedError}
            notebookService={notebookService}
            priceService={priceService}
            onClose={() => this.setState({ dialogDisplayed: false })}
            onCompletion={() => this.initializeDetailsFromServer()}
          />
        )}
      </span>
    );
  }

  private async initializeDetailsFromServer() {
    const { detailsServer, detailsService } = this.props;
    try {
      const details = (await detailsServer.getUtilizationData()) as Details;
      const [machineTypes, acceleratorTypes] = await Promise.all([
        detailsService.getMachineTypes(),
        detailsService.getAcceleratorTypes(),
      ]);

      details.machineTypes = getMachineTypeConfigurations(machineTypes);
      details.acceleratorTypes = acceleratorTypes;

      this.setState({ details });
    } catch (e) {
      console.warn('Unable to retrieve GCE VM details');
      this.setState({ receivedError: true });
    }
  }

  private async updateUtilizationData() {
    const { detailsServer } = this.props;
    try {
      const details = (await detailsServer.getUtilizationData()) as Details;

      this.setState({
        details: {
          ...this.state.details,
          utilization: details.utilization,
        },
      });
    } catch (e) {
      console.warn('Unable to retrieve GCE VM details');
      this.setState({ receivedError: true });
    }
  }

  private cycleDisplayed() {
    const [d1, d2] = this.state.displayedAttributes;
    const displayedAttributes: [number, number] = [
      (d1 + 1) % MAPPED_ATTRIBUTES.length,
      (d2 + 1) % MAPPED_ATTRIBUTES.length,
    ];
    const [a1, a2] = MAPPED_ATTRIBUTES.slice(
      displayedAttributes[0],
      displayedAttributes[1] + 1
    );
    let shouldRefresh = false;
    if (
      REFRESHABLE_MAPPED_ATTRIBUTES.includes(a1) ||
      REFRESHABLE_MAPPED_ATTRIBUTES.includes(a2)
    ) {
      shouldRefresh = true;
    }

    this.setState({
      displayedAttributes,
      shouldRefresh,
    });
  }

  private getDisplayedDetails(details: Details): JSX.Element {
    const { displayedAttributes } = this.state;
    return (
      <React.Fragment>
        {displayedAttributes.map((d, i) => (
          <span
            key={i}
            className={STYLES.attribute}
            title={`${MAPPED_ATTRIBUTES[d].label} - Click to cycle`}
            onClick={() => this.cycleDisplayed()}
          >
            {MAPPED_ATTRIBUTES[d].mapper(details)}
            {i === 0 ? ' | ' : ''}
          </span>
        ))}
      </React.Fragment>
    );
  }
}

/** Top-level widget exposed to JupyterLab for showing VM details. */
export class VmDetailsWidget extends ReactWidget {
  constructor(
    private detailsServer: ServerWrapper,
    private notebookService: NotebooksService,
    private detailsService: DetailsService,
    private priceService: PriceService,
    private initializationError: boolean
  ) {
    super();
  }

  render() {
    return (
      <VmDetails
        detailsServer={this.detailsServer}
        detailsService={this.detailsService}
        notebookService={this.notebookService}
        priceService={this.priceService}
        initializationError={this.initializationError}
      />
    );
  }
}
