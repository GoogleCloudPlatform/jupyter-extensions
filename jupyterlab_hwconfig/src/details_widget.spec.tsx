/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock('@jupyterlab/apputils');

import * as React from 'react';
import { shallow } from 'enzyme';
import { DETAILS_RESPONSE } from './test_helpers';
import { STYLES } from './data/styles';
import { VmDetails } from './details_widget';
import { HardwareConfigurationDialog } from './components/hardware_configuration_dialog';
import { HardwareService } from './service/hardware_service';
import { NotebooksService } from './service/notebooks_service';
import { PriceService } from './service/price_service';

describe('VmDetails', () => {
  const mockGetDetails = jest.fn();
  const mockHardwareService = ({
    getVmDetails: mockGetDetails,
  } as unknown) as HardwareService;
  const mockNotebookService = {} as NotebooksService;
  const mockPriceService = {} as PriceService;
  const detailsResponsePromise = Promise.resolve(JSON.parse(DETAILS_RESPONSE));

  beforeEach(() => {
    jest.resetAllMocks();
    jest.useFakeTimers();
    mockGetDetails.mockReturnValue(detailsResponsePromise);
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it('Renders with details', async () => {
    const vmDetails = shallow(
      <VmDetails
        hardwareService={mockHardwareService}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
      />
    );
    expect(vmDetails).toMatchSnapshot('Retrieving');
    await detailsResponsePromise;

    expect(vmDetails).toMatchSnapshot('Details');
    expect(mockGetDetails).toHaveBeenCalledTimes(1);
  });

  it('Renders with get details error', async () => {
    mockGetDetails.mockImplementation(() => {
      throw new Error();
    });

    const vmDetails = shallow(
      <VmDetails
        hardwareService={mockHardwareService}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
      />
    );
    expect(vmDetails).toMatchSnapshot('Received Error');

    vmDetails.find('[title="Show all details"]').simulate('click');
    vmDetails.update();

    expect(vmDetails.state('dialogDisplayed')).toEqual(true);
    expect(vmDetails.find(HardwareConfigurationDialog)).toHaveLength(1);

    const hardwareConfigurationDialog = vmDetails.find(
      HardwareConfigurationDialog
    );
    expect(hardwareConfigurationDialog.props().receivedError).toEqual(true);
  });

  it('Opens dialog when icon is clicked', async () => {
    const vmDetails = shallow(
      <VmDetails
        hardwareService={mockHardwareService}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
      />
    );
    await detailsResponsePromise;

    vmDetails.find('[title="Show all details"]').simulate('click');
    vmDetails.update();

    expect(vmDetails.state('dialogDisplayed')).toEqual(true);
    expect(vmDetails.find(HardwareConfigurationDialog)).toHaveLength(1);
  });

  it('Cycles through attributes when clicked', async () => {
    const vmDetails = shallow(
      <VmDetails
        hardwareService={mockHardwareService}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
      />
    );
    await detailsResponsePromise;

    expect(mockGetDetails).toHaveBeenCalledTimes(1);

    let attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.length).toBe(2);
    expect(attributes.first().text()).toBe('pytorch | ');
    expect(attributes.last().text()).toBe('test-project');

    attributes.first().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('test-project | ');
    expect(attributes.last().text()).toBe('PyTorch:1.4');

    attributes.first().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('PyTorch:1.4 | ');
    expect(attributes.last().text()).toBe('4 vCPU, 15 GB RAM (n1-standard-4)');

    // Simulate two clicks
    attributes.last().simulate('click');
    attributes.last().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('CPU: 50.0% | ');
    expect(attributes.last().text()).toBe('Memory: 16.0%');

    attributes.first().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('Memory: 16.0% | ');
    expect(attributes.last().text()).toBe('GPU: Tesla K80 - 100.0%');
  });

  it('Auto-refreshes when resource utilization are displayed', async () => {
    const vmDetails = shallow(
      <VmDetails
        hardwareService={mockHardwareService}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
      />
    );
    await detailsResponsePromise;

    expect(mockGetDetails).toHaveBeenCalledTimes(1);

    // Click four times to move to CPU usage
    for (let i = 0; i < 4; i++) {
      vmDetails
        .find(`span.${STYLES.attribute}`)
        .first()
        .simulate('click');
    }

    let attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('CPU: 50.0% | ');
    expect(attributes.last().text()).toBe('Memory: 16.0%');

    jest.advanceTimersToNextTimer();
    expect(mockGetDetails).toHaveBeenCalledTimes(2);

    attributes.first().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('Memory: 16.0% | ');
    expect(attributes.last().text()).toBe('GPU: Tesla K80 - 100.0%');
    jest.advanceTimersToNextTimer();
    expect(mockGetDetails).toHaveBeenCalledTimes(3);

    // Click twice and timer should not refresh data again
    attributes.first().simulate('click');
    attributes.first().simulate('click');
    attributes = vmDetails.find(`span.${STYLES.attribute}`);
    expect(attributes.first().text()).toBe('pytorch | ');
    expect(attributes.last().text()).toBe('test-project');
    jest.advanceTimersToNextTimer();
    expect(mockGetDetails).toHaveBeenCalledTimes(3);
  });
});
