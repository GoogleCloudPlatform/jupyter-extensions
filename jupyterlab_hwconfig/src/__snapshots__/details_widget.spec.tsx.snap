// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VmDetails Renders with details: Details 1`] = `
<span
  className="container_f20cyc7"
>
  <span
    className="icon_fptpv5r jp-VmStatusIcon interactiveHover_feuq7k1"
    onClick={[Function]}
    title="Show all details"
  />
  <WidgetPopup>
    <ResourceUtilizationCharts
      detailsServer={
        Object {
          "getUtilizationData": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        }
      }
    />
  </WidgetPopup>
  <span
    className="interactiveHover_feuq7k1"
  >
    <span
      className="attribute_f1pfon3k"
      key="0"
      onClick={[Function]}
      title="VM Name - Click to cycle"
    >
      pytorch
       | 
    </span>
    <span
      className="attribute_f1pfon3k"
      key="1"
      onClick={[Function]}
      title="Project - Click to cycle"
    >
      test-project
    </span>
  </span>
</span>
`;

exports[`VmDetails Renders with details: Retrieving 1`] = `
<span
  className="container_f20cyc7"
>
  <span
    className="icon_fptpv5r jp-VmStatusIcon interactiveHover_feuq7k1"
    onClick={[Function]}
    title="Show all details"
  />
  <WidgetPopup>
    <ResourceUtilizationCharts
      detailsServer={
        Object {
          "getUtilizationData": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        }
      }
    />
  </WidgetPopup>
  <span
    className="interactiveHover_feuq7k1"
  >
    Retrieving VM Details...
  </span>
</span>
`;

exports[`VmDetails Renders with get details error: Received Error 1`] = `
<span
  className="container_f20cyc7"
>
  <span
    className="icon_fptpv5r jp-VmStatusIcon interactiveHover_feuq7k1"
    onClick={[Function]}
    title="Show all details"
  />
  <WidgetPopup>
    <ResourceUtilizationCharts
      detailsServer={
        Object {
          "getUtilizationData": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "throw",
                "value": [Error],
              },
            ],
          },
        }
      }
    />
  </WidgetPopup>
  <span
    className="interactiveHover_feuq7k1"
  >
    Error retrieving VM Details
  </span>
</span>
`;
